crates/aitia/src/tests.rs

```diff
@@ -0,0 +1,476 @@
+    }
+    {
+        let (graph, _passes) = Cause::from(Countdown(3)).traverse(&true_0).fail().unwrap();
+        assert_eq!(
+            graph.nodes(),
+            maplit::hashset![
+                Cause::from(Countdown(1)),
+                Cause::from(Countdown(2)),
+                Cause::from(Countdown(3))
+            ]
+        );
+
+        assert_eq!(graph.edge_count(), 2);
+    }
+    {
+        let (graph, _passes) = Cause::from(Countdown(3)).traverse(&true_1).fail().unwrap();
+        let nodes = graph.node_weights().cloned().collect::<HashSet<_>>();
+
+        assert_eq!(
+            nodes,
+            maplit::hashset![Cause::from(Countdown(2)), Cause::from(Countdown(3))]
+        );
+
+        assert_eq!(graph.edge_count(), 1);
+    }
+}
+
+#[test]
+fn loopy() {
+    holochain_trace::test_run().ok().unwrap();
+
+    #[derive(Clone, Debug, PartialEq, Eq, Hash, derive_more::Display)]
+    struct Countdown(u8);
+
+    impl Fact for Countdown {
+        type Context = Checks<Self>;
+
+        fn cause(&self, _: &Self::Context) -> Option<Cause<Self>> {
+            match self.0 {
+                3 => Some(Self(2).into()),
+                2 => Some(Self(1).into()),
+                1 => Some(Self(3).into()),
+                0 => None,
+                _ => unreachable!(),
+            }
+        }
+
+        fn check(&self, ctx: &Self::Context) -> bool {
+            (ctx)(self)
+        }
+    }
+
+    let true_0: Checks<Countdown> = Box::new(|i| i.0 == 0);
+    let true_1: Checks<Countdown> = Box::new(|i| i.0 == 1);
+    let true_3: Checks<Countdown> = Box::new(|i| i.0 == 3);
+
+    assert!(matches!(
+        Cause::from(Countdown(3)).traverse(&true_0),
+        Traversal::Groundless
+    ));
+
+    {
+        let tr = Cause::from(Countdown(3)).traverse(&true_1);
+        let (graph, _passes) = tr.fail().unwrap();
+        assert_eq!(
+            graph.nodes(),
+            maplit::hashset![Cause::from(Countdown(3)), Cause::from(Countdown(2))]
+        );
+        assert_eq!(graph.edge_count(), 1);
+    }
+    {
+        let tr = Cause::from(Countdown(2)).traverse(&true_3);
+        let (graph, _passes) = tr.fail().unwrap();
+        assert_eq!(
+            graph.nodes(),
+            maplit::hashset![Cause::from(Countdown(2)), Cause::from(Countdown(1))]
+        );
+        assert_eq!(graph.edge_count(), 1);
+    }
+}
+
+#[test]
+fn branching_any() {
+    holochain_trace::test_run().ok().unwrap();
+
+    #[derive(Clone, Debug, PartialEq, Eq, Hash, derive_more::Display)]
+    struct Branching(u8);
+
+    impl Fact for Branching {
+        type Context = HashSet<u8>;
+
+        fn cause(&self, _ctx: &Self::Context) -> Option<Cause<Self>> {
+            (self.0 <= 64)
+                .then(|| Cause::Any(vec![Self(self.0 * 2).into(), Self(self.0 * 2 + 1).into()]))
+        }
+
+        fn check(&self, ctx: &Self::Context) -> bool {
+            ctx.contains(&self.0)
+        }
+    }
+
+    {
```

Comment:
I see there's some setup for these test cases and I'd slightly prefer modules with the required custom types and multiple tests within the modules rather than multiple tests inside one test. What would be useful though is comments describing what each of these blocks is checking. These are useful up here https://github.com/holochain/holochain/pull/2971/files#diff-325c7d393b1bd42a607544136ff50e1f1406f408d3a9d1f1e12d5d07b701f20aR61

Reply:
Check them out now, I cleaned them up with `test-case`

Reply:
Much prefer this, thanks!